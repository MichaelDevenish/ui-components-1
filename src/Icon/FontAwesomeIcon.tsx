import React from 'react'
import { Icon, IIconProps } from './Icon'

export class FontAwesomeIcon extends React.PureComponent<IIconProps> {
  public render (): JSX.Element {
    const {
      type
    } = this.props

    return (
      <Icon
        {...this.props}
        type={`fa-${type}`}
      />
    )
  }
}
