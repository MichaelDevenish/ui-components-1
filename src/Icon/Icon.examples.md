# Icon

### Basic Icon

```jsx
  <Icon
    type='star'
  />
```
### Large Icon

```jsx
  <Icon
    type='star'
    isLarge
  />
```
  
### 4x size Icon
  
```jsx
  <Icon
    type='star'
    size={4}
  />
```
  
### Spinning Icon
  
```jsx
  <Icon
    type='star'
    isSpinning
  />
```
  
### Coloured Icon
  
```jsx
  <Icon
    type='star'
    color='rgb(204,204,0)'
  />
```

### Mix and match

Icon props can be used together!

```jsx
  <Icon
    type='star'
    color='rgb(204,204,0)'
    isSpinning
    size={3}
  />
```
