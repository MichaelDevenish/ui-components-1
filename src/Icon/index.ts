export { FontAwesomeIcon } from './FontAwesomeIcon'
export { IntelliIcon } from './IntelliIcon'
