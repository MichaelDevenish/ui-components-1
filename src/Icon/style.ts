import styled, { css } from 'styled-components'

type tSize = 'xsmall' | 'small' | 'medium' | 'large' | 'xlarge' | 'xxlarge'

interface StyledIconProps {
  color?: string
  customSize?: number
  tSize?: tSize
}

interface IBadgeWrapper {
  tSize?: tSize
}

const BadgeWrapper = styled.span`
  position: absolute;
  
  ${(props: IBadgeWrapper) => {
    switch (props.tSize) {
      case 'large':
        return `
          margin-top: -10px;
          margin-left: -10px;
        `
      case 'xlarge':
        return `
          margin-top: -9px;
          margin-left: -15px;
        `
      case 'xxlarge':
        return `
          margin-top: -9px;
          margin-left: -22px;
        `
    }
  }}
`

const StyledIcon = styled.i`
  ${(props: StyledIconProps) => props.color && css`
    color: ${props.color};  
  `}
  
  ${(props: StyledIconProps) => props.customSize && css`
    font-size: ${props.customSize}rem;  
  `}
  
  ${(props: StyledIconProps) => {
    switch (props.tSize) {
      case 'xsmall':
        return `
          font-size: .75rem; // 12px
        `
      case 'small':
        return `
          font-size: .85rem; // 14px
        `
      case 'medium':
        return `
          font-size: 1rem; // 16px
        `
      case 'large':
        return `
          font-size: 1.5rem; // 24px
        `
      case 'xlarge':
        return `
          font-size: 2rem; // 32px
        `
      case 'xxlarge':
        return `
          font-size: 3rem; // 48px
        `
    }
  }}

  &.icon-small {
    bottom: 2px;
    font-size: xx-small;
    position: relative;
  }
`

export {
  BadgeWrapper,
  StyledIcon,
  StyledIconProps
}
