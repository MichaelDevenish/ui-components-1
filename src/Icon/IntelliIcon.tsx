import React from 'react'
import { Icon, IconProps } from './Icon'

export class IntelliIcon extends React.PureComponent<IconProps> {
  public render (): JSX.Element {
    const {
      type
    } = this.props

    return (
      <Icon
        {...this.props}
        type={`intelli-icon-${type}`}
      />
    )
  }
}
