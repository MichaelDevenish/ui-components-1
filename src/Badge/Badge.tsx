import React from 'react'
import { FontAwesomeIcon } from '../Icon'
import { BadgeWrapper } from './style'

interface IBadgeProps {
  /** Label of the badge  */
  label?: number | string
  /** Is the badge content loading */
  pending?: boolean
  /** Background color of the badge */
  backgroundColor?: string
  /** Font color of the badge */
  color?: string
  /** Is the badge small */
  isSmall?: boolean
  /** Is the badge large */
  isLarge?: boolean
}

class Badge extends React.PureComponent<IBadgeProps> {
  get label (): number | string | JSX.Element | undefined {
    const {
      pending,
      label
    } = this.props

    if (pending) {
      return (
        <FontAwesomeIcon
          type='refresh'
          isSpinning
          className='refresh-icon'
        />
      )
    }

    return label
  }

  public render (): JSX.Element {
    const {
      backgroundColor,
      color,
      isSmall,
      isLarge
    } = this.props

    return (
      <BadgeWrapper
        backgroundColor={backgroundColor}
        color={color}
        isSmall={isSmall}
        isLarge={isLarge}
      >
        {this.label}
      </BadgeWrapper>
    )
  }
}

export {
  Badge
}
