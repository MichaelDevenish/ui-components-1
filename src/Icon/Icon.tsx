import React from 'react'
import classNames from 'classnames'
import { StyledIcon, BadgeWrapper } from './style'

export interface IconProps {
  /** FontAwesome or alternate name of the icon to display */
  type: string
  /** Multiplies icon size by this amount (max 5) */
  size?: 1 | 2 | 3 | 4 | 5
  /** Adds FontAwesome stacked class to icon */
  isStacked?: boolean
  /** Adds FontAwesome `fa-lg` class to icon */
  isLarge?: boolean
  /** Adds a class to reduce the font size of the icon */
  isSmall?: boolean
  /** Colour of the icon */
  color?: string
  /** Additional class name to pass to the icon */
  className?: string
  /** Adds FontAwesome `fa-spin` class to icon */
  isSpinning?: boolean
  /** Badge to display on the icon */
  badge?: JSX.Element
}

export class Icon extends React.PureComponent<IconProps> {
  public static defaultProps: Partial<IconProps> = {
    isStacked: false,
    isLarge: false,
    isSmall: false,
    isSpinning: false
  }

  get sizeClass (): string {
    const {
      size,
      isStacked,
      isLarge
    } = this.props

    if (isLarge) {
      return `fa-lg`
    }

    if (size) {
      if (isStacked) {
        return `fa-stack-${size}x`
      }

      return `fa-${size}x`
    }

    return ''
  }

  get classNames (): string {
    const {
      className,
      type,
      isSpinning,
      isSmall
    } = this.props

    return classNames(
      className,
      'icon',
      'fa',
      type,
      this.sizeClass,
      {
        'fa-spin': isSpinning,
        'icon-small': isSmall
      }
    )
  }

  get badgeSize () {
    const {
      size
    } = this.props

    if (size === 3) {
      return {
        isSmall: true
      }
    }

    if (size === 5) {
      return {
        isLarge: true
      }
    }
  }

  get badge () {
    const {
      badge,
      size
    } = this.props

    if (badge && size && size >= 3) {
      return (
        <BadgeWrapper
          size={size}
        >
          {React.cloneElement(badge, {
            isSmall: size === 3,
            isLarge: size === 5
          })}
        </BadgeWrapper>
      )
    }
  }

  public render (): JSX.Element {
    const {
      color
    } = this.props

    return (
      <React.Fragment>
        <StyledIcon
          className={this.classNames}
          aria-hidden
          color={color}
        />

        {this.badge}
      </React.Fragment>
    )
  }
}
