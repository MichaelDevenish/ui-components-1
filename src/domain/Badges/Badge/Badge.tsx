import React from 'react'
import { FontAwesomeIcon } from '@Domain/Icons'
import { BadgeWrapper } from './style'

interface IBadgeProps {
  /** Label of the badge  */
  label?: number | string
  /** Is the badge content loading */
  pending?: boolean
  /** Background color of the badge */
  backgroundColor?: string
  /** Font color of the badge */
  color?: string
  /** Enforced badge guideline sizes */
  size?: 'small' | 'medium' | 'large'
}

class Badge extends React.PureComponent<IBadgeProps> {
  public static defaultProps: Partial<IBadgeProps> = {
    size: 'small'
  }

  get label (): number | string | JSX.Element | undefined {
    const {
      pending,
      label
    } = this.props

    if (pending) {
      return (
        <FontAwesomeIcon
          type='refresh'
          isSpinning={true}
          className='refresh-icon'
        />
      )
    }

    return label
  }

  public render (): JSX.Element {
    const {
      backgroundColor,
      color,
      size
    } = this.props

    return (
      <BadgeWrapper
        color={color}
        size={size}
        backgroundColor={backgroundColor}
      >
        {this.label}
      </BadgeWrapper>
    )
  }
}

export {
  Badge
}
