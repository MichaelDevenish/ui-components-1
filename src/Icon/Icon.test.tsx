import { mount } from 'enzyme'
import React from 'react'

import { Icon } from './Icon'
import {Badge} from '../Badge'

describe('<Icon />', () => {
  it('should render icon with no size or colour', () => {
    const wrapper = mount(
      <Icon
        type='fa-check'
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render icon with size', () => {
    const wrapper = mount(
      <Icon
        type='fa-check'
        size={3}
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render icon with stack and size', () => {
    const wrapper = mount(
      <Icon
        type='fa-check'
        size={2}
        isStacked
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render icon with no stack or size if only isStacked is given', () => {
    const wrapper = mount(
      <Icon
        type='fa-check'
        isStacked
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render a large icon', () => {
    const wrapper = mount(
      <Icon
        type='circle'
        isLarge
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render icon with colour', () => {
    const wrapper = mount(
      <Icon
        type='fa-check'
        color='#1A2B3C'
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render a non-FontAwesome icon', () => {
    const wrapper = mount(
      <Icon
        type='intelli-icon-avatar'
      />
    )

    expect(wrapper).toMatchSnapshot()
  })

  it('should render an icon with a badge', () => {
    const wrapper = mount(
      <Icon
        type='fa-check'
        size={3}
        badge={
          <Badge
            label={6}
            backgroundColor='#1A2B3C'
            color='#1A293C'
          />
        }
      />
    )

    expect(wrapper).toMatchSnapshot()
  })
})
