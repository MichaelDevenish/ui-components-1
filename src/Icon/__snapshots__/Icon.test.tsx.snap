// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Icon /> should render a large icon 1`] = `
.c0.icon-small {
  bottom: 2px;
  font-size: xx-small;
  position: relative;
}

<Icon
  isLarge={true}
  isSmall={false}
  isSpinning={false}
  isStacked={false}
  type="circle"
>
  <styled.i
    aria-hidden={true}
    className="icon fa circle fa-lg"
  >
    <i
      aria-hidden={true}
      className="icon fa circle fa-lg c0"
    />
  </styled.i>
</Icon>
`;

exports[`<Icon /> should render a non-FontAwesome icon 1`] = `
.c0.icon-small {
  bottom: 2px;
  font-size: xx-small;
  position: relative;
}

<Icon
  isLarge={false}
  isSmall={false}
  isSpinning={false}
  isStacked={false}
  type="intelli-icon-avatar"
>
  <styled.i
    aria-hidden={true}
    className="icon fa intelli-icon-avatar"
  >
    <i
      aria-hidden={true}
      className="icon fa intelli-icon-avatar c0"
    />
  </styled.i>
</Icon>
`;

exports[`<Icon /> should render an icon with a badge 1`] = `
.c0.icon-small {
  bottom: 2px;
  font-size: xx-small;
  position: relative;
}

<Icon
  badge={
    <Badge
      backgroundColor="#1A2B3C"
      color="#1A293C"
      label={6}
      tSize="small"
    />
  }
  isLarge={false}
  isSmall={false}
  isSpinning={false}
  isStacked={false}
  size={3}
  type="fa-check"
>
  <styled.i
    aria-hidden={true}
    className="icon fa fa-check fa-3x"
  >
    <i
      aria-hidden={true}
      className="icon fa fa-check fa-3x c0"
    />
  </styled.i>
</Icon>
`;

exports[`<Icon /> should render icon with colour 1`] = `
.c0 {
  color: #1A2B3C;
}

.c0.icon-small {
  bottom: 2px;
  font-size: xx-small;
  position: relative;
}

<Icon
  color="#1A2B3C"
  isLarge={false}
  isSmall={false}
  isSpinning={false}
  isStacked={false}
  type="fa-check"
>
  <styled.i
    aria-hidden={true}
    className="icon fa fa-check"
    color="#1A2B3C"
  >
    <i
      aria-hidden={true}
      className="icon fa fa-check c0"
      color="#1A2B3C"
    />
  </styled.i>
</Icon>
`;

exports[`<Icon /> should render icon with no size or colour 1`] = `
.c0.icon-small {
  bottom: 2px;
  font-size: xx-small;
  position: relative;
}

<Icon
  isLarge={false}
  isSmall={false}
  isSpinning={false}
  isStacked={false}
  type="fa-check"
>
  <styled.i
    aria-hidden={true}
    className="icon fa fa-check"
  >
    <i
      aria-hidden={true}
      className="icon fa fa-check c0"
    />
  </styled.i>
</Icon>
`;

exports[`<Icon /> should render icon with no stack or size if only isStacked is given 1`] = `
.c0.icon-small {
  bottom: 2px;
  font-size: xx-small;
  position: relative;
}

<Icon
  isLarge={false}
  isSmall={false}
  isSpinning={false}
  isStacked={true}
  type="fa-check"
>
  <styled.i
    aria-hidden={true}
    className="icon fa fa-check"
  >
    <i
      aria-hidden={true}
      className="icon fa fa-check c0"
    />
  </styled.i>
</Icon>
`;

exports[`<Icon /> should render icon with size 1`] = `
.c0.icon-small {
  bottom: 2px;
  font-size: xx-small;
  position: relative;
}

<Icon
  isLarge={false}
  isSmall={false}
  isSpinning={false}
  isStacked={false}
  size={3}
  type="fa-check"
>
  <styled.i
    aria-hidden={true}
    className="icon fa fa-check fa-3x"
  >
    <i
      aria-hidden={true}
      className="icon fa fa-check fa-3x c0"
    />
  </styled.i>
</Icon>
`;

exports[`<Icon /> should render icon with stack and size 1`] = `
.c0.icon-small {
  bottom: 2px;
  font-size: xx-small;
  position: relative;
}

<Icon
  isLarge={false}
  isSmall={false}
  isSpinning={false}
  isStacked={true}
  size={2}
  type="fa-check"
>
  <styled.i
    aria-hidden={true}
    className="icon fa fa-check fa-stack-2x"
  >
    <i
      aria-hidden={true}
      className="icon fa fa-check fa-stack-2x c0"
    />
  </styled.i>
</Icon>
`;
